mlir/lib/Util/Util.cpp

```diff
@@ -833,34 +834,43 @@ LogicalResult air::canonicalizeWrapAndStrideList(OpBuilder builder,
     strides.erase(strides.begin() + i);
   }
 
-  SmallVector<int> redundant_dims;
-  if (!sizes.empty())
-    for (int i = sizes.size() - 1; i >= 1; i--) {
-      if (getConstantIntValue(sizes[i]) && getConstantIntValue(sizes[i - 1]) &&
+  bool hasRedudantDims = false;
+  if (!sizes.empty()){
+    int i = sizes.size() - 1;
+    while (i >= 1) {
+      if (getConstantIntValue(offsets[i]) && getConstantIntValue(offsets[i - 1]) &&
+          getConstantIntValue(sizes[i]) && getConstantIntValue(sizes[i - 1]) &&
           getConstantIntValue(strides[i]) &&
           getConstantIntValue(strides[i - 1])) {
         auto const_size = *getConstantIntValue(sizes[i]);
         auto const_size_next = *getConstantIntValue(sizes[i - 1]);
         auto const_stride = *getConstantIntValue(strides[i]);
         auto const_stride_next = *getConstantIntValue(strides[i - 1]);
-        // Skip over the first dimension if stride is 1
-        if (const_stride == 1 && i == (int)sizes.size() - 1)
-          continue;
         if (const_stride_next == const_size * const_stride) {
-          redundant_dims.push_back(i - 1);
+          hasRedudantDims = true;
           sizes[i] = builder.create<arith::ConstantIndexOp>(
               builder.getUnknownLoc(), const_size * const_size_next);
+          offsets.erase(offsets.begin() + i - 1);
+          sizes.erase(sizes.begin() + i - 1);
+          strides.erase(strides.begin() + i - 1);
         }
       }
+      i--;
     }
+  }
 
-  for (auto i : redundant_dims) {
-    offsets.erase(offsets.begin() + i);
-    sizes.erase(sizes.begin() + i);
-    strides.erase(strides.begin() + i);
+  // If default data access pattern, then clear the offsets, sizes and strides.
+  if (offsets.size() == 1 && sizes.size() == 1 && strides.size() == 1){
+    if (getConstantIntValue(offsets[0]) && getConstantIntValue(sizes[0]) && getConstantIntValue(strides[0])){
+      if (*getConstantIntValue(strides[0]) == 1 && *getConstantIntValue(sizes[0]) == memref_volume){
+        offsets.erase(offsets.begin());
+        sizes.erase(sizes.begin());
+        strides.erase(strides.begin());
+      }
+    }
   }
 
-  if (unit_dims.empty() && redundant_dims.empty()) {
+  if (unit_dims.empty() && !hasRedudantDims) {
```

Comment:
Is it possible to erase the vectors in the loop above but not trigger these conditions? i.e. a false `failure()`?